server/images.go

```diff
@@ -300,250 +263,197 @@ func CreateModel(ctx context.Context, name, modelFileDir, quantization string, c
-					deleteMap[layer.Digest] = struct{}{}
-					if layer.MediaType == "application/vnd.ollama.image.params" {
-						fromParamsPath, err := GetBlobsPath(layer.Digest)
-						if err != nil {
-							return err
-						}
-
-						fromParamsFile, err := os.Open(fromParamsPath)
-						if err != nil {
-							return err
-						}
-						defer fromParamsFile.Close()
-
-						if err := json.NewDecoder(fromParamsFile).Decode(&fromParams); err != nil {
-							return err
-						}
-					}
+			var err2 error
+			var tempfiles []*os.File
 
-					layer, err := NewLayerFromLayer(layer.Digest, layer.MediaType, modelpath.GetShortTagname())
+			// TODO(mxyng): replace with rangefunc
+			layers2.Items()(func(layer *Layer, ggml *llm.GGML) bool {
+				if quantization != "" && ggml != nil && ggml.Name() == "gguf" {
+					ftype, err := llm.ParseFileType(quantization)
 					if err != nil {
-						return err
+						err2 = err
+						return false
 					}
 
-					layers.Add(layer)
-				}
+					filetype := ggml.KV().FileType()
+					if !slices.Contains([]string{"F16", "F32"}, filetype) {
+						err2 = errors.New("quantization is only supported for F16 and F32 models")
+						return false
+					}
 
-				deleteMap[manifest.Config.Digest] = struct{}{}
-				continue
-			}
-			defer bin.Close()
+					fn(api.ProgressResponse{Status: fmt.Sprintf("quantizing %s model to %s", filetype, quantization)})
 
-			var offset int64
-			for {
-				fn(api.ProgressResponse{Status: "creating model layer"})
-				if _, err := bin.Seek(offset, io.SeekStart); err != nil {
-					return err
-				}
+					blob, err := GetBlobsPath(layer.Digest)
+					if err != nil {
+						err2 = err
+						return false
+					}
 
-				ggml, size, err := llm.DecodeGGML(bin)
-				if errors.Is(err, io.EOF) {
-					break
-				} else if errors.Is(err, llm.ErrUnsupportedFormat) {
-					return fmt.Errorf("model binary specified in FROM field is not a valid gguf format model, %w", err)
-				} else if err != nil {
-					return err
-				}
+					temp, err := os.CreateTemp(filepath.Dir(blob), quantization)
+					if err != nil {
+						err2 = err
+						return false
+					}
+					tempfiles = append(tempfiles, temp)
 
-				config.SetModelFormat(ggml.Name())
-				config.SetModelFamily(ggml.KV().Architecture())
-				config.SetModelType(format.HumanNumber(ggml.KV().ParameterCount()))
-				config.SetFileType(ggml.KV().FileType())
+					if err := llm.Quantize(blob, temp.Name(), ftype); err != nil {
+						err2 = err
+						return false
+					}
 
-				mediatype := mediatype
-				if ggml.KV().Architecture() == "clip" {
-					mediatype = "application/vnd.ollama.image.projector"
+					layer, err = NewLayer(temp, layer.MediaType)
+					if err != nil {
+						err2 = err
+						return false
+					}
 				}
 
-				sr := io.NewSectionReader(bin, offset, size)
-				layer, err := NewLayer(sr, mediatype)
-				if err != nil {
-					return err
+				if ggml != nil {
+					config.ModelFormat = cmp.Or(config.ModelFormat, ggml.Name())
```

Comment:
These `cmp.Or()` calls end up being confusing, paricularly since nothing is set in `config` already. Can you reduce them to `config.ModelFormat = ggml.Name()`, etc? Each of the initial settings is going to be unset anyway, so even if the second argument is empty it's not going to make a difference.

Reply:
As much as I hated the setters, maybe that's the way to go? It's just not super intuitive as it's written.

Reply:
I don't agree. It's pretty intuitive to me. The documentation for [`cmp.Or`](https://pkg.go.dev/cmp#Or) is also very clear:

> Or returns the first of its arguments that is not equal to the zero value. If no argument is non-zero, it returns the zero value. 

It's similar to Python:

```python
config.model_format = config.model_format || ggml.name()
```